database/rrdcontext.c

```diff
@@ -0,0 +1,2730 @@
+    if(unlikely(!host)) return;
+    if(unlikely(!host->rrdctx)) return;
+
+    if(host->rrdctx_queue) {
+        dictionary_destroy((DICTIONARY *)host->rrdctx_queue);
+        host->rrdctx_queue = NULL;
+    }
+
+    dictionary_destroy((DICTIONARY *)host->rrdctx);
+    host->rrdctx = NULL;
+}
+
+static inline bool rrdcontext_should_be_deleted(RRDCONTEXT *rc) {
+    if(likely(!(rc->flags & RRD_FLAG_DELETED)))
+        return false;
+
+    if(likely(!(rc->flags & RRD_FLAG_LIVE_RETENTION)))
+        return false;
+
+    if(unlikely(rc->flags & RRD_FLAGS_PREVENTING_DELETIONS))
+        return false;
+
+    if(unlikely(dictionary_stats_referenced_items(rc->rrdinstances) != 0))
+        return false;
+
+    if(unlikely(dictionary_stats_entries(rc->rrdinstances) != 0))
+        return false;
+
+    if(unlikely(rc->first_time_t || rc->last_time_t))
+        return false;
+
+    return true;
+}
+
+static void rrdcontext_trigger_updates(RRDCONTEXT *rc, bool force, RRD_FLAGS reason) {
+    if(unlikely(!force && !(rc->flags & RRD_FLAG_UPDATED))) return;
+
+    rrdcontext_lock(rc);
+    rc->flags |= reason;
+    rrd_flag_unset_updated(rc);
+
+    size_t min_priority = LONG_MAX;
+    RRD_FLAGS combined_instances_flags = RRD_FLAG_NONE;
+    time_t min_first_time_t = LONG_MAX, max_last_time_t = 0;
+    size_t instances_active = 0, instances_deleted = 0;
+    {
+        RRDINSTANCE *ri;
+        dfe_start_read(rc->rrdinstances, ri) {
+            // find the combined flags of all the metrics
+            combined_instances_flags |= ri->flags & RRD_FLAGS_PROPAGATED_UPSTREAM;
+
+            if (unlikely(rrdinstance_should_be_deleted(ri))) {
+                instances_deleted++;
+                continue;
+            }
+
+            instances_active++;
+
+            if (ri->priority > 0 && ri->priority < min_priority)
+                min_priority = ri->priority;
+
+            if (ri->first_time_t && ri->first_time_t < min_first_time_t)
+                min_first_time_t = ri->first_time_t;
+
+            if (ri->last_time_t && ri->last_time_t > max_last_time_t)
+                max_last_time_t = ri->last_time_t;
+        }
+        dfe_done(ri);
+    }
+
+    rc->flags &= ~RRD_FLAG_DELETED;
+
+    if(unlikely(!instances_active && instances_deleted)) {
+        // we had some instances, but they are gone now...
+
+        rc->flags |= RRD_FLAG_DELETED;
+        rrd_flag_set_updated(rc, RRD_FLAG_UPDATE_REASON_ZERO_RETENTION);
+    }
+    else if(instances_active) {
+        // we have some active instances...
+
+        if (unlikely(min_first_time_t == LONG_MAX))
+            min_first_time_t = 0;
+
+        if (unlikely(min_first_time_t == 0 && max_last_time_t == 0)) {
+            rc->first_time_t = 0;
+            rc->last_time_t = 0;
+
+            if(unlikely(combined_instances_flags & RRD_FLAG_LIVE_RETENTION)) {
+                rc->flags |= RRD_FLAG_DELETED;
+                rrd_flag_set_updated(rc, RRD_FLAG_UPDATE_REASON_ZERO_RETENTION);
+            }
```

Comment:
They will marked as deleted, but they will not be deleted if the NEW_OBJECT and UPDATED_OBJECT flags are set. Check the `should_be_deleted()` functions.

Reply:
The point is that we should mark them as deleted, because we don't have any other hooks to actually let the cloud know. Once they are marked as deleted, the will be queued to be sent to the cloud, with a delay of 1 minute. If within one minute they are collected, they will be deduped. Otherwise they will be deleted and recreated once they are collected.